echo_server.py

```diff
@@ -40,46 +63,106 @@ def parse_request(header):
+    if len(first_line) < 3:
+        # For requests that are invalid due missing information:
+        raise ValueError(400, 'Bad Request')
+    # When .accept() is interupted to kill the server, method will not have
+    # been bound at the point the script continues after .accept() is forced
+    # to return, UnboundLocalError needs to be handled where parse_request()
+    # is being called.
+    elif method != 'GET' or proto != 'HTTP/1.1':
         # Deny non-GET and non-HTTP/1.1 requests.
         raise ValueError(403, 'Forbidden')
 
-    # Divide headers by line
-    try:
-        headers = []
-        for line in header[1:]:
-            headers.append(line)
-    except IndexError:
-        pass
-    return uri, headers
+    return uri
 
 
-def response_ok():
+def response_ok(content_type=None, body=None):
     response = """\
 HTTP/1.1 200 OK\r\n\
-"""
+Date: {date}\r\n\
+Content-type: {content_type}\r\n\
+Content-length: {content_length}\r\n\
+\r\n\
+{body}\r\n\
+""".format(date=datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT'),
+           content_type=content_type,
+           content_length=len(str(body)),
+           body=body)
     return response
 
 
 def response_error(e):
     response = """\
 HTTP/1.1 {type} {cause}\r\n\
-""".format(type=e[0], cause=e[1])
+Date: {date}\r\n\
+\r\n\
+{type} {cause}\r\n\
+""".format(type=e[0],
+           cause=e[1],
+           date=datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT'))
     return response
 
 
-if __name__ == '__main__':
-    buffsize = 8
+def resolve_uri(uri):
+    # Form uri into one useable for finding things
+    parsed = urlparse(uri)
+    req_directory, file_name = os.path.split(parsed.path)
+    print req_directory
+    print file_name
+    # Relative file path
+    req_directory = os.path.abspath(req_directory[1:])
+    print req_directory
+
+    file_location = os.path.join(req_directory, file_name)
+    print file_location
+    target, ext = os.path.splitext(file_location)
+    print target
+    print ext
+
+    if not os.path.exists(file_location):
+        raise ValueError(404, 'File Not Found')
+
+    content_type = None
+    if ext == '.html':
+        content_type = 'text/html'
+        with open(file_location) as f:
+            body = f.read()
+    elif ext == '.txt' or ext == '.py':
+        content_type = 'text/plain'
+        with open(file_location) as f:
+            body = f.read()
+    elif ext == '.jpg' or ext == '.png':
+        content_type = 'image/gif'
+        body = urllib.urlopen(urllib.pathname2url(file_location)).read()
+    elif not ext:
+        content_type = 'text/html'
+        body = '<h1>{}</h1>'.format(os.path.split(target)[1])
+        for item in os.listdir(target):
+            body = '{}<p>{}</p>'.format(
+                body, item)
+            # body = '{}<p><a href="{}">{}</a></p>'.format(
+            #     body, os.path.join(target, item), item)
+    else:
+        raise ValueError(404, 'File Not Found')
```

Comment:
It feels to me like you started this process assuming that the uri would point at a file, and then adjusted everything from there. It'd be better to use something like `os.path.isfile` or `os.path.isdir`.  In addition, you want to use the mimetypes module instead of doing your own file extension mapping, since there's so may other file extensions that are possible than you could ever build `elif` classes to cover.